src/lib/app.js

```diff
@@ -71,49 +71,50 @@ if (serviceWorkerIsSupported(window.location.hostname)) {
+  // included via importScripts as well as the SW itself. (This works from mid-2018 everywhere, but
+  // it seems sane to check.)
   const allowedHostnames = [
     'web.dev',
     'web-dev-staging.appspot.com',
     'localhost',
   ];
   return (
     'serviceWorker' in navigator &&
+    'updateViaCache' in ServiceWorkerRegistration.prototype &&
     (allowedHostnames.includes(hostname) || hostname.endsWith('.netlify.com'))
   );
 }
 
 function ensureServiceWorker() {
-  const ensurePartialCache = (isFirstInstall = false) => {
-    const {pathname} = window.location;
-    if (isFirstInstall) {
-      // We don't fetch the partial for the initial, real, HTML fetch from out HTTP server. This
-      // ensures that if the user goes offline and reloads for some reason, the page still loads.
-      getPartial(pathname);
-    }
-    if (pathname !== '/') {
-      // Aggressively refetch the landing page every time the site is loaded.
-      // TODO(samthor): Check Workbox's cache time and fetch if needed. Additionally, cache a
-      // number of recent articles.
-      getPartial('/');
-    }
-  };
-
+  const {pathname} = window.location;
   const isFirstInstall = !navigator.serviceWorker.controller;
   if (isFirstInstall) {
     // Watch for the brand new Service Worker to be activated. We claim this foreground page
     // inside the Service Worker, so this event will fire when it is activated.
-    navigator.serviceWorker.addEventListener('controllerchange', () => {
-      ensurePartialCache(true);
-    });
-  } else {
-    // This isn't the first install, but ensure some partials are up-to-date.
-    ensurePartialCache();
+    navigator.serviceWorker.addEventListener(
+      'controllerchange',
+      (event) => {
+        // We don't fetch the partial for the initial, real, HTML fetch from our HTTP server. This
+        // ensures that if the user goes offline and reloads for some reason, the page still loads.
+        getHTML(pathname);
 
-    // We claim active clients if the Service Worker's architecture rev changes. We can't
-    // reliably force a reload via the Client interface as it's unsupported in Safari.
-    navigator.serviceWorker.addEventListener('controllerchange', () => {
-      window.location.reload();
-    });
+        // Don't reload on first claim if this is the first install.
+        event.stopImmediatePropagation();
+      },
+      {once: true},
+    );
+  } else if (pathname !== '/') {
+    // Aggressively refetch the landing page every time the site is loaded.
+    // TODO(samthor): Check Workbox's cache time and fetch if needed. Additionally, cache a
+    // number of recent articles.
+    getHTML('/');
   }
-  navigator.serviceWorker.register('/sw.js');
+
+  // We claim active clients if the Service Worker's architecture rev changes. We can't
+  // reliably force a reload via the Client interface as it's unsupported in Safari.
+  navigator.serviceWorker.addEventListener('controllerchange', () => {
```

Comment:
does `controllerchange` fire if the page goes from having no service worker to having one? Just want to make sure we don't accidentally reload folks the first time they visit the site.

Reply:
The MDN example seemed to update that it could happen for new or updated SWs:

```
// Then, register a handler to detect when a new or
  // updated service worker takes control.
  navigator.serviceWorker.oncontrollerchange = function() {
    console.log('This page is now controlled by:', navigator.serviceWorker.controller);
  };
```

https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer

Reply:
Testing this out locally I haven't seen it actually fire on a new page that doesn't have a sw yet but it might be worth testing a bit more, especially in other browsers in case their implementation is different.

Reply:
I've seen the first `controllerchange` fire, but the second will be rare as it needs us to bump the major version inside the SW code. There's two cases:

1. The site is open, and I open a new page via SPA; its resources version is wrong, so I reload the site—the old SW still catches the reload, but that normal reload causes a SW check itself, and if its major version has changed then we'll reload _again_.

2. I open the site and we get a new SW which claims immediately, because of the resources version. It reloads due to this handler.